src/fu-engine.c

```diff
@@ -3050,6 +3145,55 @@ fu_engine_load_hwids (FuEngine *self)
 		g_warning ("Failed to load HWIDs: %s", error->message);
 }
 
+static gboolean
+fu_engine_update_pending_database (FuEngine *self, GError **error)
+{
+	g_autoptr(GPtrArray) devices_pending = NULL;
+
+	/* get any pending devices */
+	devices_pending = fu_pending_get_devices (self->pending, error);
+	if (devices_pending == NULL)
+		return FALSE;
+	for (guint i = 0; i < devices_pending->len; i++) {
+		FuDevice *dev_pending = g_ptr_array_index (devices_pending, i);
+		FuDevice *dev;
+		FwupdRelease *rel_pending;
+
+		/* not in the required state */
+		if (fu_device_get_update_state (dev_pending) != FWUPD_UPDATE_STATE_NEEDS_REBOOT)
+			continue;
+
+		/* is in the device list */
+		dev = fu_device_list_find_by_id (self->device_list,
+						 fu_device_get_id (dev_pending),
+						 NULL);
+		if (dev == NULL) {
+			g_debug ("no device %s found", fu_device_get_id (dev_pending));
+			continue;
+		}
+
+		/* does the installed version match what we tried to install
+		 * before fwupd was restarted */
+		rel_pending = fu_device_get_release_default (dev_pending);
+		if (rel_pending == NULL)
+			continue;
+		if (g_strcmp0 (fu_device_get_version (dev),
+			       fwupd_release_get_version (rel_pending)) != 0) {
+			g_debug ("installed version %s not matching pending %s",
+				 fu_device_get_version (dev),
+				 fwupd_release_get_version (rel_pending));
+			continue;
```

Comment:
Don't you need to set this to FWUPD_UPDATE_STATE_FAILED here?

Reply:
I'm just working on that bit, we actually have to call the `fu_plugin_get_results` vfunc to get the error message set too.